frame/multisig/src/lib.rs

```diff
@@ -485,12 +509,126 @@ pub mod pallet {
+		///
+		/// NOTE: This extrinsic is only intended to be used for creating a multisig with an expiry.
+		/// It should not be used for approving existing multisigs.
+		///
+		/// ## Complexity
+		/// - `O(S)`.
+		/// - Up to one balance-reserve or unreserve operation.
+		/// - One passthrough operation, one insert, both `O(S)` where `S` is the number of
+		///   signatories. `S` is capped by `MaxSignatories`, with weight being proportional.
+		/// - One encode & hash, both of complexity `O(S)`.
+		/// - Up to one binary search and insert (`O(logS + S)`).
+		/// - I/O: 1 read `O(S)`, up to 1 mutate `O(S)`. Up to one remove.
+		/// - One event.
+		/// - Storage: inserts one item, value size bounded by `MaxSignatories`, with a deposit
+		///   taken for its lifetime of `DepositBase + threshold * DepositFactor`.
+		#[pallet::call_index(4)]
+		#[pallet::weight({
+			let s = other_signatories.len() as u32;
+
+			// The weight is the same as for `approve_as_multi` but there is an
+			// additional write since we are writing into `MultisigExpiries`. 
+			T::WeightInfo::approve_as_multi_create(s)
+				.max(T::WeightInfo::approve_as_multi_approve(s))
+				.saturating_add(*max_weight)
+				.saturating_add(T::DbWeight::get().writes(1))
+		})]
+		pub fn create_multisig_with_expiry(
+			origin: OriginFor<T>,
+			threshold: u16,
+			other_signatories: Vec<T::AccountId>,
+			maybe_timepoint: Option<Timepoint<BlockNumberFor<T>>>,
+			call_hash: [u8; 32],
+			max_weight: Weight,
+			expiry: BlockNumberFor<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			Self::operate(
+				who,
+				threshold,
+				other_signatories,
+				maybe_timepoint,
+				CallOrHash::Hash(call_hash),
+				max_weight,
+				Some(expiry),
+			)
+		}
+
+		/// Clear up all the state of an expired multisig operation. Any deposit reserved previously
+		/// for this operation will be unreserved on success.
+		///
+		/// Anyone can call this and free up the not needed state of the expired multisig.
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// - `threshold`: The total number of approvals for this dispatch before it is executed.
+		/// - `signatories`: The accounts who can approve this dispatch.
+		/// - `timepoint`: The timepoint (block number and transaction index) of the first approval
+		/// transaction for this dispatch.
+		/// - `call_hash`: The hash of the call to be executed.
+		///
+		/// ## Complexity
+		/// - `O(S)`.
+		/// - Up to one balance-reserve or unreserve operation.
+		/// - One passthrough operation, one insert, both `O(S)` where `S` is the number of
+		///   signatories. `S` is capped by `MaxSignatories`, with weight being proportional.
+		/// - One encode & hash, both of complexity `O(S)`.
+		/// - One event.
+		/// - I/O: 1 read `O(S)`, one remove.
+		/// - Storage: removes one item.
+		#[pallet::call_index(5)]
+		#[pallet::weight(T::WeightInfo::cancel_as_multi(signatories.len() as u32))]
+		pub fn clear_multi(
```

Comment:
Why do we need a new method here to clean up instead of automatically cleaning up inside `operate` when
- Multisig executed successfully
- Multisig is found to be invalid, e.g. due to being expired

or when `cancel_as_multi` executes successfully?

Reply:
Because this does not cover all cases. In a scenario where someone creates a multisig operation that expires since everyone forgot about it, we want to allow anyone to clean it up.

Reply:
Ah I see  what you mean by 'everyone forgot about it'. What about updating `cancel_as_multi` so it can be called by anyone if the multisig is expired?

Reply:
The `cancel_as_multi` is used to cancel an operation when at least a certain threshold of the multisig parties agree to it. 
In essence, what you are saying is possible, we could allow the multisig to be canceled by anyone if it is expired but it would just look weird in my opinion since the caller would need to provide all the other arguments that do not affect his operation of clearing an expired multisig.

Reply:
> The cancel_as_multi is used to cancel an operation when at least a certain threshold of the multisig parties agree to it.

`cancel_as_multi` as I understand allows *the creator* of a multisig to clear it from storage to unlock their deposit.

IMO we should leave it that way, just expand it to also clear the expiry and refund the associated deposit if it exists. 

There is never any incentive for anyone except the creator to clean up their multisig, only the creator will get the deposit back.

The chain is protected from spam because of the deposit.

I just don't see enough benefit from new method to justify making the (already complex) API more complex.

Reply:
> cancel_as_multi as I understand allows the creator of a multisig to clear it from storage to unlock their deposit.

Seems like I interpreted this function incorrectly. In this case, I agree that we can do it inside `cancel_as_multi`

Reply:
> The chain is protected from spam because of the deposit.

This isn't really about spam. While I partially agree with you I still have the opinion that anyone should be able to clean up the blockchain state.
The thing with expired mutlisig operations is that after their expiry they are no good anymore. There is no use in storing the data onchain. We shouldn't disallow people from cleaning up the blockchain state IMO.

However, I totally agree that the naming was pretty bad. Having `cancel_as_multi` and `clear_multi` extrinsics seems super weird.

For this reason, I refactored the code a little bit. I did the following things:
- I renamed `clear_multi` into `clear_expired_multi`. The new name should be pretty self-explanatory now. 
- I moved the code responsible for clearing up a mutlisig operation into a new helper function. This is now called from both `cancel_as_multi` and `clear_expired_multi`.
- As you suggested I am now removing the multisig state inside the `operate` function. I am also using the new helper function here so that there is no code repetition.

Would be great if you could review my refactored code and lmk what you think about it.

Reply:
> I renamed clear_multi into clear_expired_multi. The new name should be pretty self-explanatory now.

I think this is a good compromise :)